File: app/src/main/java/com/example/learningverbs/firebase/CustomOnCompleteListener.java
package com.example.learningverbs.firebase;

import androidx.annotation.NonNull;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;

public abstract class CustomOnCompleteListener<T> implements OnCompleteListener<T> {
    public CustomOnCompleteListener() {
        showLoaging();
    }

    @Override
    public void onComplete(@NonNull Task<T> task) {
        hideLoading();

        if (task.isSuccessful()) {
            onSuccess(task.getResult());

        } else {
            onFailure(task.getException());

        }


    }

    public abstract void onSuccess(T task);

    public abstract void onFailure(Throwable throwable);

    public abstract void showLoaging();

    public abstract void hideLoading();

}

File: app/src/main/java/com/karla/learningverbs/kotlin/utils/firebase/CustomOnCompleteListener.kt
package com.karla.learningverbs.kotlin.utils.firebase

import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task

abstract class CustomOnCompleteListener<T> : OnCompleteListener<T> {
    init {
        showLoaging()
    }

    override fun onComplete(task: Task<T>) {
        hideLoading()
        if (task != null && task.isSuccessful) {
            onSuccess(task.result)
        } else {
            onFailure(task.exception)
        }
    }

    abstract fun onSuccess(task: T)
    abstract fun onFailure(throwable: Throwable?)
    abstract fun showLoaging()
    abstract fun hideLoading()
}